x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java

```diff
@@ -849,14 +869,91 @@ private DataType[] allowedEnrichTypes(String matchType) {
         UnresolvedAttribute ua = new UnresolvedAttribute(up.source(), up.pattern());
-        Predicate<Attribute> matcher = a -> up.match(a.name());
-        var matches = AnalyzerRules.maybeResolveAgainstList(matcher, () -> ua, attrList, true, a -> Analyzer.handleSpecialFields(ua, a));
-        return potentialCandidatesIfNoMatchesFound(ua, matches, attrList, list -> UnresolvedNamePattern.errorMessage(up.pattern(), list));
+        return resolveAgainstList(
+            a -> up.match(a.name()),
+            null,
+            ua,
+            attrList,
+            true,
+            list -> UnresolvedNamePattern.errorMessage(up.pattern(), list)
+        );
     }
 
     private static List<Attribute> resolveAgainstList(UnresolvedAttribute ua, Collection<Attribute> attrList) {
-        var matches = AnalyzerRules.maybeResolveAgainstList(ua, attrList, a -> Analyzer.handleSpecialFields(ua, a));
-        return potentialCandidatesIfNoMatchesFound(ua, matches, attrList, list -> UnresolvedAttribute.errorMessage(ua.name(), list));
+        String suffix = "." + ua.name();
+        Predicate<Attribute> matchLoosely = a -> a.name().equals(ua.name()) || a.name().endsWith(suffix);
+        Predicate<Attribute> matchExactly = a -> a.name().equals(ua.name());
+
+        return resolveAgainstList(
+            matchLoosely,
+            matchExactly,
+            ua,
+            attrList,
+            false,
+            list -> UnresolvedAttribute.errorMessage(ua.name(), list)
+        );
+    }
+
+    private static List<Attribute> resolveAgainstList(
+        Predicate<Attribute> matchLoosely,
+        @Nullable Predicate<Attribute> matchExact,
+        UnresolvedAttribute ua,
+        Collection<Attribute> attrList,
+        boolean isPattern,
+        Function<List<String>, String> messageProducer
+    ) {
+        List<Attribute> matches = maybeResolveAgainstList(matchLoosely, matchExact, ua, attrList, isPattern);
+        return potentialCandidatesIfNoMatchesFound(ua, matches, attrList, messageProducer);
+    }
+
+    /**
+     * Take an {@link UnresolvedAttribute} and find matching attributes in a given collection of attributes.
+     * Throws {@link IllegalArgumentException} if we fail to resolve.
+     * <p>
+     * If {@code matchExactly} is not {@code null}, it will be used to narrow down multiple matches found with {@code matchLoosely}.
+     * This allows predicates that match a field name {@code b} to a qualified field {@code a.b} if there's no other fields ending in
+     * {@code .b} (loose match), while still matching {@code b} correctly if there's both fields {@code a.b} and {@code b} present
+     * (exact match).
+     * <p>
+     * Regardless of {@code matchLoosely} and {@code matchExactly}, {@link Attribute}s for which {@link Attribute#synthetic()} is true are
+     * never resolved against.
+     */
+    private static List<Attribute> maybeResolveAgainstList(
```

Comment:
Moved here from `AnalyzerRules` and refactored a little to suit this PR's needs.